File: services/dev-env-service/src/halt.ts
export type HaltReason = "kill_switch" | "step_failure" | "project_switch" | "user_stop";

export type Boundary = "immediate" | "between_actions" | "between_tasks";

export const BOUNDARY_MEANING: Record<Boundary, string> = {
  immediate: "the process tree is killed without waiting",
  between_actions: "the action in flight completes, then nothing further starts",
  between_tasks: "the task in flight completes, then nothing further starts",
};

const REASONS: readonly HaltReason[] = [
  "kill_switch",
  "step_failure",
  "project_switch",
  "user_stop",
];

export interface HaltSignal {
  id: string;
  reason: HaltReason;
  request_id: string | null;
  project_id: string | null;
  task_id: string | null;
  issued_at: string;
}

export interface HaltAcknowledgement {
  service: string;
  boundary: Boundary;
  boundary_meaning: string;
  halted: boolean;
  signal_id: string;
  reason: HaltReason;
  accepted_at: string;
  signals_received: number;
  already_halted: boolean;
}

export interface HaltStatus {
  service: string;
  boundary: Boundary;
  boundary_meaning: string;
  halted: boolean;
  signal: HaltSignal | null;
  signals_received: number;
  halted_at: string | null;
  resumed_at: string | null;
  resumed_by: string | null;
}

export class HaltedError extends Error {
  constructor(
    readonly service: string,
    readonly signal: HaltSignal,
  ) {
    super(`${service} is halted by ${signal.reason} and accepts no new work`);
    this.name = "HaltedError";
  }
}

export function parseSignal(body: unknown): HaltSignal {
  const candidate = body as Partial<Record<keyof HaltSignal, unknown>> | null;

  const id = candidate?.id;
  const reason = candidate?.reason;
  const issuedAt = candidate?.issued_at;

  if (typeof id !== "string" || id.length === 0) {
    throw new TypeError("halt signal requires a non-empty id");
  }

  if (typeof reason !== "string" || !REASONS.includes(reason as HaltReason)) {
    throw new TypeError(`unknown halt reason "${String(reason)}"`);
  }

  if (typeof issuedAt !== "string" || Number.isNaN(Date.parse(issuedAt))) {
    throw new TypeError("halt signal requires an issued_at timestamp");
  }

  return {
    id,
    reason: reason as HaltReason,
    request_id: typeof candidate?.request_id === "string" ? candidate.request_id : null,
    project_id: typeof candidate?.project_id === "string" ? candidate.project_id : null,
    task_id: typeof candidate?.task_id === "string" ? candidate.task_id : null,
    issued_at: issuedAt,
  };
}

export class HaltState {
  private readonly signals: HaltSignal[] = [];
  private stopped = false;
  private haltedAt: string | null = null;
  private resumedAt: string | null = null;
  private resumedBy: string | null = null;

  constructor(
    private readonly service: string,
    private readonly boundary: Boundary,
  ) {}

  get halted(): boolean {
    return this.stopped;
  }

  get signal(): HaltSignal | null {
    return this.signals.at(-1) ?? null;
  }

  accept(signal: HaltSignal): HaltAcknowledgement {
    const already = this.stopped;
    const now = new Date().toISOString();

    this.signals.push(signal);
    this.stopped = true;

    if (!already) {
      this.haltedAt = now;
      this.resumedAt = null;
      this.resumedBy = null;
    }

    return {
      service: this.service,
      boundary: this.boundary,
      boundary_meaning: BOUNDARY_MEANING[this.boundary],
      halted: true,
      signal_id: signal.id,
      reason: signal.reason,
      accepted_at: now,
      signals_received: this.signals.length,
      already_halted: already,
    };
  }

  resume(actor: string): HaltStatus {
    this.stopped = false;
    this.resumedAt = new Date().toISOString();
    this.resumedBy = actor;

    return this.status();
  }

  guard(): void {
    const current = this.signals.at(-1);

    if (this.stopped && current) {
      throw new HaltedError(this.service, current);
    }
  }

  status(): HaltStatus {
    return {
      service: this.service,
      boundary: this.boundary,
      boundary_meaning: BOUNDARY_MEANING[this.boundary],
      halted: this.stopped,
      signal: this.signal,
      signals_received: this.signals.length,
      halted_at: this.haltedAt,
      resumed_at: this.resumedAt,
      resumed_by: this.resumedBy,
    };
  }
}

File: packages/contracts-py/src/bison_contracts/halt.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

HaltReason = Literal["kill_switch", "step_failure", "project_switch", "user_stop"]

Boundary = Literal["immediate", "between_actions", "between_tasks"]

BOUNDARY_MEANING: dict[Boundary, str] = {
    "immediate": "the process tree is killed without waiting",
    "between_actions": "the action in flight completes, then nothing further starts",
    "between_tasks": "the task in flight completes, then nothing further starts",
}


class HaltSignal(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str = Field(min_length=1)
    reason: HaltReason
    request_id: str | None = None
    project_id: str | None = None
    task_id: str | None = None
    issued_at: datetime


class HaltAcknowledgement(BaseModel):
    service: str
    boundary: Boundary
    boundary_meaning: str
    halted: bool
    signal_id: str
    reason: HaltReason
    accepted_at: datetime
    signals_received: int
    already_halted: bool


class HaltStatus(BaseModel):
    service: str
    boundary: Boundary
    boundary_meaning: str
    halted: bool
    signal: HaltSignal | None
    signals_received: int
    halted_at: datetime | None
    resumed_at: datetime | None
    resumed_by: str | None


class HaltedError(RuntimeError):
    def __init__(self, service: str, signal: HaltSignal) -> None:
        super().__init__(f"{service} is halted by {signal.reason} and accepts no new work")
        self.service = service
        self.signal = signal


class HaltState:
    def __init__(self, service: str, boundary: Boundary) -> None:
        self._service = service
        self._boundary = boundary
        self._signals: list[HaltSignal] = []
        self._halted = False
        self._halted_at: datetime | None = None
        self._resumed_at: datetime | None = None
        self._resumed_by: str | None = None

    @property
    def service(self) -> str:
        return self._service

    @property
    def boundary(self) -> Boundary:
        return self._boundary

    @property
    def halted(self) -> bool:
        return self._halted

    @property
    def signal(self) -> HaltSignal | None:
        return self._signals[-1] if self._signals else None

    def accept(self, signal: HaltSignal) -> HaltAcknowledgement:
        already = self._halted
        now = datetime.now(UTC)

        self._signals.append(signal)
        self._halted = True

        if not already:
            self._halted_at = now
            self._resumed_at = None
            self._resumed_by = None

        return HaltAcknowledgement(
            service=self._service,
            boundary=self._boundary,
            boundary_meaning=BOUNDARY_MEANING[self._boundary],
            halted=True,
            signal_id=signal.id,
            reason=signal.reason,
            accepted_at=now,
            signals_received=len(self._signals),
            already_halted=already,
        )

    def resume(self, actor: str) -> HaltStatus:
        self._halted = False
        self._resumed_at = datetime.now(UTC)
        self._resumed_by = actor

        return self.status()

    def guard(self) -> None:
        if self._halted:
            signal = self._signals[-1]
            raise HaltedError(self._service, signal)

    def status(self) -> HaltStatus:
        return HaltStatus(
            service=self._service,
            boundary=self._boundary,
            boundary_meaning=BOUNDARY_MEANING[self._boundary],
            halted=self._halted,
            signal=self.signal,
            signals_received=len(self._signals),
            halted_at=self._halted_at,
            resumed_at=self._resumed_at,
            resumed_by=self._resumed_by,
        )
